[Security] Document `access_token.token_handler.oidc_user_info` (#17463)
symfony/symfony-docs: security/access_token.rst

    ``failure_handler`` option and create a class that implements
    :class:`Symfony\\Component\\Security\\Http\\Authentication\\AuthenticationFailureHandlerInterface`.

> Configure the OidcUserInfoTokenHandler

Reviewer: Should we add an additional headline "Using OIDC" with two sub headlines for the token handlers?

                https://symfony.com/schema/dic/security/security-1.0.xsd">

            <config>
                <firewall name="main">
                    <access-token>
                        <token-handler>
                            <oidc-user-info>
                                <client base-uri="https://www.example.com/realms/demo/protocol/openid-connect/userinfo"/>
                            </oidc-user-info>
                        </token-handler>
                    </access-token>
                </firewall>
            </config>
        </srv:container>

    .. code-block:: php

        // config/packages/security.php
        use Symfony\Config\SecurityConfig;

        return static function (SecurityConfig $security) {
            $security->firewall('main')
                ->accessToken()
                    ->tokenHandler()
                        ->oidcUserInfo()
                            ->client()
                                ->baseUri('https://www.example.com/realms/demo/protocol/openid-connect/userinfo')
            ;
        };

.. tip::

    Following the `OpenID Connect Specification`_, the `sub` claim

>     is used by default as user identifier. To use another claim,

Reviewer: ```suggestion
    is used as user identifier by default. To use another claim,
```

            <config>
                <firewall name="main">
                    <access-token>
                        <token-handler>
                            <oidc-user-info claim="email">
                                <client base-uri="https://www.example.com/realms/demo/protocol/openid-connect/userinfo"/>
                            </oidc-user-info>
                        </token-handler>
                    </access-token>
                </firewall>
            </config>
        </srv:container>

    .. code-block:: php

        // config/packages/security.php
        use Symfony\Config\SecurityConfig;

        return static function (SecurityConfig $security) {
            $security->firewall('main')
                ->accessToken()
                    ->tokenHandler()
                        ->oidcUserInfo()
                            ->claim('email')
                            ->client()
                                ->baseUri('https://www.example.com/realms/demo/protocol/openid-connect/userinfo')
            ;
        };

.. tip::

    The `oidc_user_info` token handler automatically creates

>     an HTTP Client with the specified configuration. If you

Reviewer: ```suggestion
    an HTTP client with the specified configuration. If you
```

>     prefer using your own client, just specify the service

Reviewer: ```suggestion
    prefer using your own client, you can specify the service
```

>     name as `client` option:

Reviewer: ```suggestion
    name via the `client` option:
```

                https://symfony.com/schema/dic/services/services-1.0.xsd
                http://symfony.com/schema/dic/security
                https://symfony.com/schema/dic/security/security-1.0.xsd">

            <config>
                <firewall name="main">
                    <access-token>
                        <token-handler>
                            <oidc-user-info client="oidc.client"/>
                        </token-handler>
                    </access-token>
                </firewall>
            </config>
        </srv:container>

    .. code-block:: php

        // config/packages/security.php
        use Symfony\Config\SecurityConfig;

        return static function (SecurityConfig $security) {
            $security->firewall('main')
                ->accessToken()
                    ->tokenHandler()
                        ->oidcUserInfo()
                            ->client('oidc.client')
            ;
        };

By default, the OidcUserInfoTokenHandler creates an OidcUser with the claims. To create your own User from the claims,
you must :doc:`create your own UserProvider </security/user_providers>`:

.. configuration-block::

>     .. code-block:: php

Reviewer: ```suggestion
you must :doc:`create your own UserProvider </security/user_providers>`::
```

            <config>
                <firewall name="main">
                    <access-token>
                        <token-handler>
                            <oidc audience="My audience">
                                <signature algorithm="HS256" key="{'kty':'...','k':'...'}"/>
                            </oidc>
                        </token-handler>
                    </access-token>
                </firewall>
            </config>
        </srv:container>

    .. code-block:: php

        // config/packages/security.php
        use Symfony\Config\SecurityConfig;

        return static function (SecurityConfig $security) {
            $security->firewall('main')
                ->accessToken()
                    ->tokenHandler()
                        ->oidc()
                            ->audience('My audience')
                            ->signature()
                                ->algorithm('HS256')
                                ->key('{"kty":"...","k":"..."}')
            ;
        };

By default, the OidcTokenHandler creates an OidcUser with the claims. To create your own User from the claims,

> you must :doc:`create your own UserProvider </security/user_providers>`:
> 
> .. configuration-block::
> 
>     .. code-block:: php

Reviewer: ```suggestion
you must :doc:`create your own UserProvider </security/user_providers>`::
```

                http://symfony.com/schema/dic/security
                https://symfony.com/schema/dic/security/security-1.0.xsd">

            <config>
                <firewall name="main">
                    <access-token>
                        <token-handler>
                            <oidc-user-info claim="email">
                                <client base-uri="https://www.example.com/realms/demo/protocol/openid-connect/userinfo"/>
                            </oidc-user-info>
                        </token-handler>
                    </access-token>
                </firewall>
            </config>
        </srv:container>

    .. code-block:: php

        // config/packages/security.php
        use Symfony\Config\SecurityConfig;

        return static function (SecurityConfig $security) {
            $security->firewall('main')
                ->accessToken()
                    ->tokenHandler()
                        ->oidcUserInfo()
                            ->claim('email')
                            ->client()
                                ->baseUri('https://www.example.com/realms/demo/protocol/openid-connect/userinfo')
            ;
        };

.. tip::

>     The `oidc_user_info` token handler automatically creates

Reviewer: ```suggestion
    The ``oidc_user_info`` token handler automatically creates
```

    an HTTP client with the specified configuration. If you
    prefer using your own client, you can specify the service

>     name via the `client` option:

Reviewer: ```suggestion
    name via the ``client`` option:
```

            xmlns:srv="http://symfony.com/schema/dic/services"
            xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
            xsi:schemaLocation="http://symfony.com/schema/dic/services
                https://symfony.com/schema/dic/services/services-1.0.xsd
                http://symfony.com/schema/dic/security
                https://symfony.com/schema/dic/security/security-1.0.xsd">

            <config>
                <firewall name="main">
                    <access-token>
                        <token-handler>
                            <oidc-user-info client="oidc.client"/>
                        </token-handler>
                    </access-token>
                </firewall>
            </config>
        </srv:container>

    .. code-block:: php

        // config/packages/security.php
        use Symfony\Config\SecurityConfig;

        return static function (SecurityConfig $security) {
            $security->firewall('main')
                ->accessToken()
                    ->tokenHandler()
                        ->oidcUserInfo()
                            ->client('oidc.client')
            ;
        };

> By default, the OidcUserInfoTokenHandler creates an OidcUser with the claims. To create your own User from the claims,

Reviewer: ```suggestion
By default, the ``OidcUserInfoTokenHandler`` creates an `OidcUser` with the claims. To create your own User from the claims,
```

you must :doc:`create your own UserProvider </security/user_providers>`::

        // src/Security/Core/User/OidcUserProvider.php

>         class OidcUserProvider implements ClaimsBasedUserProviderInterface

Reviewer: Please add the needed use statements

        {
            public function loadUserByIdentifier(string $identifier, array $claims = []): UserInterface
            {
                // do some magic
            }

>         }

Reviewer: Only 4 spaces for indentation please

                http://symfony.com/schema/dic/security
                https://symfony.com/schema/dic/security/security-1.0.xsd">

            <config>
                <firewall name="main">
                    <access-token>
                        <token-handler>
                            <oidc claim="email">
                                <signature algorithm="HS256" key="{'kty':'...','k':'...'}"/>
                            </oidc>
                        </token-handler>
                    </access-token>
                </firewall>
            </config>
        </srv:container>

    .. code-block:: php

        // config/packages/security.php
        use Symfony\Config\SecurityConfig;

        return static function (SecurityConfig $security) {
            $security->firewall('main')
                ->accessToken()
                    ->tokenHandler()
                        ->oidc()
                            ->claim('email')
                            ->signature()
                                ->algorithm('HS256')
                                ->key('{"kty":"...","k":"..."}')
            ;
        };

.. tip::

>     The `oidc` token handler also check for the token audience.

Reviewer: ```suggestion
    The ``oidc`` token handler also check for the token audience.
```

    By default, this audience is optional. To enable this check,

>     add the `audience` option:

Reviewer: ```suggestion
    add the ``audience`` option:
```

                        <token-handler>
                            <oidc audience="My audience">
                                <signature algorithm="HS256" key="{'kty':'...','k':'...'}"/>
                            </oidc>
                        </token-handler>
                    </access-token>
                </firewall>
            </config>
        </srv:container>

    .. code-block:: php

        // config/packages/security.php
        use Symfony\Config\SecurityConfig;

        return static function (SecurityConfig $security) {
            $security->firewall('main')
                ->accessToken()
                    ->tokenHandler()
                        ->oidc()
                            ->audience('My audience')
                            ->signature()
                                ->algorithm('HS256')
                                ->key('{"kty":"...","k":"..."}')
            ;
        };

By default, the OidcTokenHandler creates an OidcUser with the claims. To create your own User from the claims,
you must :doc:`create your own UserProvider </security/user_providers>`::

        // src/Security/Core/User/OidcUserProvider.php

>         class OidcUserProvider implements ClaimsBasedUserProviderInterface

Reviewer: Please add the needed use statements

        {
            public function loadUserByIdentifier(string $identifier, array $claims = []): UserInterface
            {
                // do some magic
            }

>         }

Reviewer: Only 4 spaces for indentation please

    ``failure_handler`` option and create a class that implements
    :class:`Symfony\\Component\\Security\\Http\\Authentication\\AuthenticationFailureHandlerInterface`.

Using OIDC
----------

> 

Reviewer: Please mention, that it is introduced as `experimental`

    ``failure_handler`` option and create a class that implements
    :class:`Symfony\\Component\\Security\\Http\\Authentication\\AuthenticationFailureHandlerInterface`.

> Using OIDC

Reviewer 2: I think using the full protocol name (OpenID Connect) would be good here, either in parenthesis or vice versa

----------

> 

Reviewer 2: It would be nice to have a short paragraph describing what OIDC is, eventually mentioning a few identity/access providers such as Keycloack, especially for people that not yet familiar with the protocol.

Author: I'll add a short paragraph introducing OIDC, but I don't think it's the purpose of this doc to present some providers or go deeper in this protocol.
